common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java

```diff
@@ -19,23 +19,64 @@
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
 import java.nio.channels.ReadableByteChannel;
 
 import io.netty.buffer.ByteBuf;
 
 public class ByteArrayReadableChannel implements ReadableByteChannel {
   private ByteBuf data;
+  private boolean closed;
 
   public int readableBytes() {
-    return data.readableBytes();
+    return data == null ? 0 : data.readableBytes();
   }
 
   public void feedData(ByteBuf buf) {
-    data = buf;
+    if (closed) {
+      buf.release();
+      return;
+    }
+    ByteBuf currentData = data;
+    if (currentData == null) {
+      data = buf;
+    } else {
+      int currentReadable = currentData.readableBytes();
```

Comment:
It seems you're making this change so that it's safe to call `feedData` when there is still data to process.

That's unnecessary; this class is not general purpose, and is only used in one place, where the surrounding code makes sure all this data is consumed after `feedData` is called. If you're worried, a check for the buffer being empty + throwing an exception should be enough.

(I'd kinda like to rewrite this stuff at some point to avoid so much copying and add new features, but that requires time...)

Reply:
@vanzin the class is public and as it is today very error-prone. What happens if someone start using it from another place ? So either we should throw or change the class to be safe to use (which I did here). If you prefer throwing that's fine as well.

Reply:
It's not a public API for Spark. Spark has a lot of public classes that are not considered part of its public API. People shouldn't use those except at their own risk.

(Only the classes in Spark's public API docs are considered really public, and this whole module is not there.)